[MOS-307] Application and Service example and docs (#3663)
mudita/MuditaOS: module-sys/README.md

   * [Switching power mode](#switching-power-mode)
   * [Power off the system](#power-off-the-system)
   * [Features missing or to be considered](#features-missing-or-to-be-considered)
      * [Dynamic clock frequency scaling during normal operation](#dynamic-clock-frequency-scaling-during-normal-operation)
      * [More sophisticated mechanism of entering and exiting from low power mode](#more-sophisticated-mechanism-of-entering-and-exiting-from-low-power-mode)
      * [Core voltage in low power mode](#core-voltage-in-low-power-mode)
* [Caveats and good practices](#caveats-and-good-practices)
   * [nullptr returns](#nullptr-returns)
   * [blocking requests](#blocking-requests)
   * [Just to not use DataReceivedHandler to handle messages, use connect(...) insead](#just-to-not-use-datareceivedhandler-to-handle-messages-use-connect-insead)
   * [DataReceivedHandler should not block for long period of time.](#datareceivedhandler-should-not-block-for-long-period-of-time)
   * [Watch out for async incoming messages](#watch-out-for-async-incoming-messages)

# Preface

The whole MuditaOS system can, with great simplification, be seen as a connection between:

> * BSP - board support layer. The low level integration code parts which are resposible for the hardware operations and abstractions.

Reviewer: ```suggestion
* BSP - board support layer. The low-level integration code parts are responsible for the hardware operations and abstractions.
```

* Workers - simple threads connecting BSP layer to Services layer. Responsible for handling the hardware abstraction.
* [Services](#Services)
    * [Bus](#Bus) connected threads
    * operating on workers to perform hardware tasks
    * communicating with other services
    * notifying applications via broadcast/multicasts & [actions](../module-apps/ModuleApps.md#System-actions)
    * Managed via [SystemManager](#System-manager)
* Applications

>     * Bus connected threads

Reviewer 2: ```suggestion
    * Bus-connected threads
```

>     * Services derivatives which have access to UI system and are spawned only for specyfic time.

Reviewer: ```suggestion
    * Services derivatives that have access to UI system and are spawned only for a specific time.
```

    * Managed via ApplicationManager

> 
> **NOTE** main difference between application and service is that application has acess to UI, while services don't and applications are working only the required time, while services are always dormant when inactive.

Reviewer: ```suggestion

**NOTE** main difference between application and service is that application has access to UI, while services don't and applications are working only the required time, while services are always dormant when inactive.
```

**WARNING:** Services are not started at the moment their constructor is called. This essentially means that `sys::Bus` and every functionality based on bus is not available yet.
The first moment when you are capable to call `sys::Bus` is `InitHandler()` of application. This is the place where your service initialization code should happen.

Then you have to implement several virtual methods:

### InitHandler()

This handler is invoked upon creation of the service.
**NOTE** It is very important to return proper return code specified in: `enum class ReturnCodes{...};`

### DeinitHandler()

Invoked upon destruction of the service. Should free allocated resources.

> * **NOTE** This is the place where Service should assure that it's workers are closed.

Reviewer: ```suggestion
* **NOTE** This is the place where Service should assure that its workers are closed.
```

**WARNING:** Please mind that safe usage of the bus is available after `InitHandler()` was called

The Bus subsystem was developed in order to allow cross-services communication.

> Prefered method to achieve service -> application communication is either:

Reviewer: ```suggestion
Preferred method to achieve service -> application communication is either:
```

- request and initialize communication from application to service
- use [actions](../module-apps/ModuleApps.md#System-actions)

The Bus enables us to:
* send asynchronous unicast messages
* send synchronous unicast messages

> * send mutlicast messages to subscribed recievers
> * send broadcast messages to all reciepients - please use with great caution

Reviewer: ```suggestion
* send multicast messages to subscribed receivers
* send broadcast messages to all recipients - please use with great caution
```

The bus communication can be split into two different parts:
* System communication - defined in SystemMessage handlers
     * Meant to provide raw system functionalities like:
        * System setup
        * System shutdown
        * System timers
    * **NOT MEANT** to provide user features like i.e. modes or custom device notifications
* Userspace communication - defined in `connect(...)`s `DataReceivedHandler()` etc
    * meant to provide an easy way to handle preferably asynchronous communication

> **All services and applications do is essentially act and optionally respond to messages on bus**. There is literally no other way to perform any action in the system properly programatically.

Reviewer: ```suggestion
**All services and applications do is essentially act and optionally respond to messages on bus**. There is literally no other way to perform any action in the system properly programmatically.
```

There are a few ways to handle messages on the bus:

> * `connect(...)` and `disconnect(...)` meant to provide an signal -> slot interaction. These handlers can be attached anywehre in the Servcie/App

Reviewer: ```suggestion
* `connect(...)` and `disconnect(...)` meant to provide an signal -> slot interaction. These handlers can be attached anywhere in the Servcie/App
```

> * `async_call(...)` -> `sync(...)` meant to provice minimal, one time request, async capabilities. As we do not have std::promise it's a poors man implementation of such capabilities in the system.

Reviewer: ```suggestion
* `async_call(...)` -> `sync(...)` meant to provide minimal, one time request, async capabilities. As we do not have std::promise it's a poor man implementation of such capabilities in the system.
```

> * `async_call(...)` -> `sync(...)` meant to provide minimal, one time request, async capabilities. As we do not have std::promise it's a poor man implementation of such capabilities in the system.

Reviewer 2: ```suggestion
* `async_call(...)` -> `sync(...)` meant to provide minimal, one time request, async capabilities. As we do not have `std::promise` it's a poor man implementation of such capabilities in the system.
```

};
```

> In current implementation only `FullSpeed` and `LowPowerIdle` modes are used. It is worth to note that `LowPowerIdle` is heavy customize, and it absolutely doesn't correspond to low power idle mode from RT1051's data sheet. Main differences are:

Reviewer 3: ```suggestion
In current implementation only `FullSpeed` and `LowPowerIdle` modes are used. It is worth to note that `LowPowerIdle` is heavy customized, and it absolutely doesn't correspond to low power idle mode from RT1051's data sheet. Main differences are:
```

Core voltage is set in LPM_EnterLowPowerIdle function which can be found in module-bsp/board/rt1051/common/clock_config.cpp.
For more info please check RT1051 Reference Manual, Chapter 18 "DCDC Converter" and DCDC Register 3.

# Caveats and good practices

## nullptr returns

If you ever set empty shared pointer to the response on message - the handling won't respond to caller
**This has very serious consequences**:
* you can't call `async_call` to it - as there will be no response
* you can't call blocking bus unicasts to it - as there will be no response

**NOTE**: This might render your application/service useless in some scenarions
**NOTE**: you can see system messages handled by services unlocking flag in [debug.hpp](../module-utils/log/api/log/debug.hpp)

## blocking requests

Blocking requests have very serious aftermath.
* You are blocking your thread till your request is handled
* Some other task can be in worst case scenario blocked on the request from which you sent blocking request
* In worst case scenario presented it can cause lock till handlers timeout or watchdog intervention

> ## Just to not use DataReceivedHandler to handle messages, use `connect(...)` insead

Reviewer 3: ```suggestion
## Just to not use DataReceivedHandler to handle messages, use `connect(...)` instead
```

## DataReceivedHandler should not block for long period of time.  
It is okay to do some minor blocking tasks but blocking for longer periods will make whole system unpredictable or behave
in very uncontrolled manner. It is perfectly fine to use `Bus::SendUnicast` blocking API in the handler.

## Watch out for async incoming messages  
Some time ago second parameter was added to DataReceivedHandler 
`DataReceivedHandler(DataMessage* msg, ResponseMessage* resp)`
This addition was dictated by need of using non-blocking variant of `Bus::SendUnicast` which then can trigger receiving async
response. In this case second parameter is not nullptr used and should contain valid `ResponseMessage`, otherwise 
`ResponseMessage* resp` will be set to nullptr. Additionally, by design there won't be the situation where both `DataMessage* msg`
and `ResponseMessage* resp` are not nullptr. Users of services should be aware of that and always check if params are valid 
before using them.

## appmgr is not started with other services - there can't be dependecy to it

> This creates a very specyfic situation where we can't depend on if service appmgr exist and is able to launch any application or service.

Reviewer: ```suggestion
This creates a very specific situation where we can't depend on if service appmgr exists and is able to launch any application or service.
```

> This creates a very specific situation where we can't depend on if service appmgr exists and is able to launch any application or service.

Reviewer 2: ```suggestion
This creates a very specific situation where we can't depend on the appmgr service existence and is able to launch any application or service.
```
